application/src/main/kotlin/net/corda/v5/application/services/TokenClaimCriteria.kt

```diff
@@ -0,0 +1,28 @@
+package net.corda.v5.application.services
+
+import net.corda.v5.base.types.MemberX500Name
+import net.corda.v5.crypto.SecureHash
+import java.math.BigDecimal
+
+/**
+ * [TokenClaimCriteria] describes the selection criteria for a token selection query using the [TokenCache] API.
+ *
+ * @property tokenType The type of tokens to be selected.
+ * @property issuerHash The [SecureHash] of issuer of tokens to be selected.
+ * @property notaryX500Name The [MemberX500Name] of the notary of tokens to be selected.
+ * @property symbol The symbol of the notary of tokens to be selected.
+ * @property targetAmount The minimum value for the sum of [ClaimedToken.amount] for the selected tokens.
+ * @property tagRegex Optional regular expression to match against the [ClaimedToken.tag] field. Null matches all.
+ * @property ownerHash Optional owner [SecureHash] of the tokens to be selected . Null matches all.
+ */
+class TokenClaimCriteria(
```

Comment:
Should we have a builder to make it easier for java users?

Why are some properties not in the constructor? Even if they're optional they should be in the constructor with another constructor provided that sets them.

Reply:
To set them in java/kotlin is pretty clean, would a builder add anything here?
```
        val criteria = TokenClaimCriteria(
            "Currency",
            getIssuerHash(),
            getNotaryX500Name(),
            "USD",
            BigDecimal(100)
        ).apply {
            // Set optional criteria
            ownerHash = getOwnerHash()
            tagRegex = "(test)"
        }
```
```
        TokenClaimCriteria criteria = new TokenClaimCriteria(
                "Currency",
                getIssuerHash(),
                getNotaryX500Name(),
                "USD",
                new BigDecimal(100)
        );

        // Set optional criteria
        criteria.setOwnerHash(getOwnerHash());
        criteria.setTagRegex("(test)");
```

Reply:
There are quite a few properties, and even in the kotlin version, without named arguments, it's not super easy to know what each input represents. I don't mind too much for kotlin since we can always used named args, but java doesn't have the ability.

Reply:
That being said, I'm really not keen on the optional args and would prefer another constructor that sets both and/or either of them.